src/lidar_subscriber.cpp

```diff
@@ -0,0 +1,156 @@
+  average.range_m = ((count-1)*existing_landmark.range_m+new_measurement.range_m)/count;
+  if(existing_landmark.bearing_rad>M_PI){
+	  existing_landmark.bearing_rad -=2*M_PI;
+  }
+  if(new_measurement.bearing_rad>M_PI){
+	  new_measurement.bearing_rad-=2*M_PI;
+  }
+  average.bearing_rad = ((count-1)*existing_landmark.bearing_rad+new_measurement.bearing_rad)/count;
+  return average;
+}
+
+Observation2D merge(Observation2D &first, Observation2D &second, const int &count1, const int &count2){
+  //standard weighted average --> between first and last landmark
+  struct Observation2D average;
+  float weight1 = ((float)count1)/((float)(count1+count2));
+  float weight2 = ((float)count2)/((float)(count1+count2));
+  if(first.bearing_rad>M_PI){
+          first.bearing_rad -=2*M_PI;
+  }
+  if(second.bearing_rad>M_PI){
+          second.bearing_rad-=2*M_PI;
+  }
+  average.range_m = first.range_m*weight1+second.range_m*weight2;
+  average.bearing_rad = first.bearing_rad*weight1+(second.bearing_rad)*weight2;
+  //std::cout<<first.bearing_rad<<" " <<second.bearing_rad<<"\n";
+  return average;
+}
+
+class LidarSubscriber : public rclcpp::Node
+{
+  public:
+    LidarSubscriber()
+   : Node("lidar_subscriber")
+    {
+      subscription = this->create_subscription<sensor_msgs::msg::LaserScan>(
+      "scan_filtered", 10, std::bind(&LidarSubscriber::laserscan_to_landmarks, this, std::placeholders::_1));
+      publisher = this->create_publisher<visualization_msgs::msg::Marker>("landmark_visualization",10);
+    }
+  private:
+    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr subscription;
+    rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr publisher;
+    std::queue<Observation2D> laserscan_to_landmarks(const sensor_msgs::msg::LaserScan::SharedPtr msg){
+        float angle_increment = msg->angle_increment;
+        int measurement_count = static_cast<int>(std::round((msg->angle_max-msg->angle_min)/angle_increment)) + 1;
+        std::queue<Observation2D> landmarks;
+        std::queue<int> counts; 
+	struct Observation2D previous_measurement;
+	struct Observation2D first_measurement;
+        for (int i = 1; i < measurement_count; i++){
+            float curr_range = msg->ranges[i];
+            if (curr_range == std::numeric_limits<float>::infinity()||curr_range == -1*std::numeric_limits<float>::infinity()){ 
+           continue;
+	    }else{
+                float curr_bearing = i * angle_increment;
+		struct Observation2D curr_measurement = {curr_range, curr_bearing};
+	       if (landmarks.empty()){
+		       landmarks.push(curr_measurement);
+		       previous_measurement = curr_measurement;
+		       first_measurement = curr_measurement;
+		       counts.push(1);	       
+	       }else if(same_landmark(curr_measurement, previous_measurement)){
+                  counts.back()++;
+                  struct Observation2D updated_landmark = merge(landmarks.back(), curr_measurement, counts.back());
```

Comment:
`merge` will probably get called a lot during the entire operation. If `merge` keeps returning `struct Observation2D` by value, the extra copy assignment in your code may slow down performance. Instead of returning the average result by value, pass in a reference to the `struct Observation2D` you are trying to modify, and modify it with in the `merge` function. In that case when you are trying to merge the latest landmark, call the `merge` function and pass in the back of the queue.